### PR TITLE
feat(genai-util): add support for core invocations

### PR DESCRIPTION
**Pull request by psx95**
## Which problem is this PR solving?

Re: #3681

#### Implemented so far in genai-utils:
The `@opentelemetry/genai-util` package provides generic lifecycle scaffolding — `TelemetryHandler` and the abstract `BaseInvocation` — but nothing that actually knows what a GenAI operation is. 

#### The current problem
There's no concrete invocation for the operations the GenAI semantic conventions define, so every instrumentation (OpenAI, Bedrock, Vertex, ...) still has to hand-assemble the per-operation semantics itself: which `gen_ai.*` attributes apply to a chat vs. an embeddings call, correct span naming (chat {model}, embeddings {model}, execute_tool {name}), span kind, request-option mapping, response model/ID and finish reasons, token usage and time-to-first-chunk for streaming, tool call arguments/results gated on content-capture mode, and matching metric attribute sets for duration and token usage.

#### What this PR adds: 
This PR adds the concrete operation layer on top of that scaffolding — `InferenceInvocation`, `EmbeddingInvocation`, and `ToolInvocation` — plus `startInference()` / `startEmbedding()` / `startTool()` factories on the handler, so instrumentation authors get spec-correct, consistent GenAI spans and metrics out of the box instead of re-deriving the conventions per package.

## Short description of the changes

- **Inference Invocation (`src/invocations/inference.ts`)**: Added `InferenceInvocation` subclass of `BaseInvocation` handling chat and text completion spans, model request/response attributes, finish reasons, token usage metrics, TTFT metrics, and content capture modes.
- **Embedding Invocation (`src/invocations/embedding.ts`)**: Added `EmbeddingInvocation` subclass of `BaseInvocation` handling embedding vector creation spans (`SpanKind.CLIENT`), server coordinates, and input token metrics.
- **Tool Invocation (`src/invocations/tool.ts`)**: Added `ToolInvocation` subclass of `BaseInvocation` handling tool execution spans (`SpanKind.INTERNAL`), tool parameters, safe circular reference serialization, and tool output capture.
- **Handler Factory Methods (`src/handler.ts`)**: Added `startInference()`, `startEmbedding()`, and `startTool()` to `TelemetryHandler`.
- **Span Name Utility (`src/utils.ts`)**: Added `getSpanName()` helper formatting `<operation.name> <request.model>` per GenAI semantic conventions.
- **Unit Tests**: Added test suites covering inference lifecycle (`test/invocations/inference.test.ts`), embedding operations (`test/invocations/embedding.test.ts`), and tool execution edge cases (`test/invocations/tool.test.ts`).

## Important Info
 - This is the **fourth PR** in a series that attempt to breakdown PR https://github.com/open-telemetry/opentelemetry-js-contrib/pull/3677. 
 - Based off: #3757

### CODE REVIEW

File: packages/genai-util/src/invocations/inference.ts [hunk 1]
```diff
@@ -0,0 +1,396 @@
...
+      metricAttrs[ATTR_GEN_AI_RESPONSE_MODEL] = this._responseModel;
+    }
+    if (this._serverAddress) {
+      metricAttrs[ATTR_SERVER_ADDRESS] = this._serverAddress;
+    }
+    if (this._serverPort !== undefined) {
+      metricAttrs[ATTR_SERVER_PORT] = this._serverPort;
+    }
+    if (error) {
+      metricAttrs[ATTR_ERROR_TYPE] = getErrorType(error);
```
Comment (JacksonWeber) [inline packages/genai-util/src/invocations/inference.ts]: The base class already passes the resolved error-type string. Calling `getErrorType()` again turns it into `_OTHER`, losing the actual error classification. You can use the string directly in both inference and embedding.

File: packages/genai-util/src/invocations/inference.ts [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+      );
+    }
+    return this;
+  }
+
+  /**
+   * Build the metric attributes shared by all metrics recorded for this invocation.
+   */
+  private _getMetricAttributes(error?: unknown): Attributes {
+    const metricAttrs: Attributes = {
```
Comment (JacksonWeber) [inline packages/genai-util/src/invocations/inference.ts:321]: Please include `this._metricAttributes` here and in embedding. Without it, attributes added through `setMetricAttribute()` or `setMetricAttributes()` are silently dropped.

File: packages/genai-util/src/invocations/inference.ts (hunk 1 above)
Comment (psx95) [reply packages/genai-util/src/invocations/inference.ts]: Good catch, the overridden `_recordMetrics` should be accepting a resolved `errorType: string` - it's unknown here.

Addressed in [2fb4413](https://github.com/open-telemetry/opentelemetry-js-contrib/pull/3762/commits/2fb44132e7da4c49c589a0973460dd87dc82203a).

File: packages/genai-util/src/invocations/inference.ts (hunk 2 above)
Comment (psx95) [reply packages/genai-util/src/invocations/inference.ts:321]: Good catch, thanks! 👍🏻 

Addressed in [2fb4413](https://github.com/open-telemetry/opentelemetry-js-contrib/pull/3762/commits/2fb44132e7da4c49c589a0973460dd87dc82203a).

**Review by JacksonWeber — APPROVED** [review]

File: packages/genai-util/src/invocations/embedding.ts [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+    return this._responseModel;
+  }
+
+  public setDimensionCount(count: number): this {
+    this._dimensionCount = count;
+    this._span.setAttribute(ATTR_GEN_AI_EMBEDDINGS_DIMENSION_COUNT, count);
+    return this;
+  }
+
+  public getDimensionCount(): number | undefined {
```
Comment (aabmass) [inline packages/genai-util/src/invocations/embedding.ts:133]: do we need getters?

wondering if we can just set stuff in attributes and forego the private member variables. also InferenceInvocation doesn't have getters

File: packages/genai-util/src/invocations/inference.ts [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+  }
+
+  if (options.serverAddress) {
+    attrs[ATTR_SERVER_ADDRESS] = options.serverAddress;
+    if (options.serverPort !== undefined) {
+      attrs[ATTR_SERVER_PORT] = options.serverPort;
+    }
+  }
+
+  Object.assign(attrs, getRequestOptionsAttributes(options.requestOptions));
```
Comment (aabmass) [inline packages/genai-util/src/invocations/inference.ts:81]: nit I noticed throughout it's a bit inconsistent on using spread `...attrs` vs `Object.assign()`

consider just assigning this on L63 as well, unless you expect it to override the attributes set above.

File: packages/genai-util/src/invocations/inference.ts [hunk 5]
```diff
@@ -0,0 +1,372 @@
...
+    if (options.systemInstructions) {
+      const formatted = formatSystemInstructions(options.systemInstructions);
+      if (formatted) {
+        attrs[ATTR_GEN_AI_SYSTEM_INSTRUCTIONS] = formatted;
+      }
+    }
+
+    if (options.inputMessages && options.inputMessages.length > 0) {
+      const formatted = formatInputMessages(options.inputMessages);
+      if (formatted) {
```
Comment (aabmass) [inline packages/genai-util/src/invocations/inference.ts:93]: nit Do we need this check too? There's also the check on L91 if the inputMessages are empty or not

File: packages/genai-util/src/handler.ts [hunk 6]
```diff
@@ -262,4 +304,29 @@ export class TelemetryHandler {
...
+   *   measurement. Pass the invocation's context explicitly, since the
+   *   measurement is often recorded after the invocation's context is no
+   *   longer active. Defaults to the currently active context.
+   */
+  public recordExecuteToolDuration(
+    durationSeconds: number,
+    attributes?: Attributes,
+    context?: Context
+  ): void {
+    if (durationSeconds < 0 || !isFinite(durationSeconds)) {
```
Comment (aabmass) [inline packages/genai-util/src/handler.ts:323]:
```suggestion
    if (durationSeconds < 0) {
```

covered by the SDK

**Review by aabmass — APPROVED** [review]

File: packages/genai-util/src/invocations/embedding.ts (hunk 3 above)
Comment (psx95) [reply packages/genai-util/src/invocations/embedding.ts]: Hmm,  yes - I can remove the getters. There's no use of them at the moment and they can always be added later.

File: packages/genai-util/src/invocations/inference.ts (hunk 5 above)
Comment (psx95) [reply packages/genai-util/src/invocations/inference.ts]: `formatInputMessages` can technically return `string | undefined` . So this check guards against the case where formatting returns undefined (in case of an error during stringify op).

File: packages/genai-util/src/invocations/inference.ts (hunk 4 above)
Comment (psx95) [reply packages/genai-util/src/invocations/inference.ts]: Addressed in [230cdd1](https://github.com/open-telemetry/opentelemetry-js-contrib/pull/3762/commits/230cdd1a269c4fb09515c3d9edcaaf80cfa9f41b).

File: packages/genai-util/src/handler.ts (hunk 6 above)
Comment (psx95) [reply packages/genai-util/src/handler.ts]: Addressed in [230cdd1](https://github.com/open-telemetry/opentelemetry-js-contrib/pull/3762/commits/230cdd1a269c4fb09515c3d9edcaaf80cfa9f41b).

**Comment by psx95** [comment]
Note: This PR should ideally be merged **after https://github.com/open-telemetry/opentelemetry-js-contrib/pull/3772 is merged** since this PR will require a rebase.
